File: MATLAB/tools/makeNPC.m
function posnpc=makeNPC(args)
arguments
    args.pos=[0,0,0];
    args.R=50;
    args.copynumber=32;
    args.dz=50;
    args.rotation=0; % in plane rotation
    args.twistangle=pi/32;
    args.shiftangle=pi/16;
    args.dR=3;
end
dphi=pi/4;
phi=0:dphi:2*pi-dphi-args.rotation; 
v0=0*phi;
switch args.copynumber
    case 8
        phiall=phi;
        zall=v0;
        Rall=v0+args.R;
    case 16
        phiall=horzcat(phi,-phi+args.twistangle);
        zall=horzcat(v0+args.dz/2,v0-args.dz/2);
        Rall=horzcat(v0,v0)+args.R;
    case 32
        phiall=horzcat(phi,phi+args.shiftangle,-phi+args.twistangle,-phi+args.twistangle-args.shiftangle);
        zall=horzcat(v0+args.dz/2,v0+args.dz/2,v0-args.dz/2,v0-args.dz/2);
        Rall=horzcat(v0+args.R+args.dR,v0+args.R,v0+args.R+args.dR,v0+args.R);
    otherwise
        disp('NPC copy number: 8, 16 or 32')
        pos=[];
        return
end

posnpc(:,1)= Rall.*cos(phiall); posnpc(:,2)= Rall.*sin(phiall);
posnpc(:,3)= zall;
posnpc=posnpc+args.pos;

File: python/tools/makeNPC.py
import numpy as np

def makeNPC(pos=[0,0,0], R=50, copynumber=32, dz=50, rotation=0, twistangle=np.pi/32, shiftangle=np.pi/16, dR=3):
    if not isinstance(pos, np.ndarray):
        pos = np.array(pos, dtype=float)

    dphi = np.pi/4
    maxphi = 2 * np.pi - dphi - rotation
    nphi = int(maxphi / dphi)
    phi = np.arange(nphi+1)*dphi
    v0 = np.zeros_like(phi)

    if copynumber == 8:
        phiall = phi
        zall = v0
        Rall = v0 + R
    elif copynumber == 16:
        phiall = np.hstack([phi, -phi+twistangle])
        zall = np.hstack([v0+dz/2, v0-dz/2])
        Rall = np.hstack([v0,v0]) + R
    elif copynumber == 32:
        phiall = np.hstack([phi, phi+shiftangle, -phi+twistangle, -phi + twistangle - shiftangle])
        zall = np.hstack([v0+dz/2, v0+dz/2, v0-dz/2, v0-dz/2])
        Rall = np.hstack([v0+R+dR, v0+R, v0+R+dR, v0+R])
    else:
        raise ValueError("NPC copy number: 8, 16 or 32")

    posnpc = np.vstack([Rall*np.cos(phiall), Rall*np.sin(phiall), zall]).T
    posnpc += pos

    return posnpc
